### ISSUE TITLE
[BUG] RKE2 private registry setting doesn't apply to rancher-system-agent

### ISSUE THREAD

**Issue by FrankYang0529** [harvester/harvester#2125]
### Update (2022/05/04)

I found that when we create an RKE2 cluster with a private registry setting, Rancher doesn't set the private registry setting to `/etc/rancher/agent/registries.yaml` or `/etc/rancher/rke2/registries.yaml`. Rancher only sets it to RKE2 `system-default-registry` setting.

For rancher-system-agent, Rancher doesn't use the private registry to get the instruction image. Rancher get the image from `SystemAgentInstallerImage` setting.
Code Reference:
* https://github.com/rancher/rancher/blob/054b5d195f115b91a5284e85270403ba6f3ac99b/pkg/provisioningv2/rke2/planner/instructions.go#L17-L20
* https://github.com/rancher/rancher/blob/054b5d195f115b91a5284e85270403ba6f3ac99b/pkg/provisioningv2/rke2/planner/planner.go#L899-L904

---
**Describe the bug**
When users want to deploy RKE2 in an air-gapped environment, they have to provide a private registry. However, rancher-system-agent will auto-reset registries content in `/etc/rancher/agent/registries.yaml`, so it can't get images from the right source.

**To Reproduce**
1. Create a Harvester cluster. (We just want to test the rancher-system-agent reset part, so it is okay to deploy the harvester in a non-air-gapped environment.)
2. Import Harvester to Rancher.
3. Create an RKE2 cluster with the following User Data.
```
write_files:
  - content: >-
      bWlycm9yczoKICBkb2NrZXIuaW86CiAgICBlbmRwb2ludDoKICAgICAgLSAiaHR0cHM6Ly9teXJl
      Z2lzdHJ5LmxvY2FsOjUwMDAiCmNvbmZpZ3M6CiAgIm15cmVnaXN0cnkubG9jYWw6NTAwMCI6CiAg
      ICB0bHM6CiAgICAgIGluc2VjdXJlX3NraXBfdmVyaWZ5OiB0cnVlCg==
    encoding: gzip+b64
    path: /etc/rancher/agent/registries.yaml
    permissions: '0644'
```
4. Login the VM and run `sudo cat /etc/rancher/agent/registries.yaml`. The content will be `{"configs":{},"mirrors":null}`, not our content.
5. Try to replace `/etc/rancher/agent/registries.yaml` with following content.
```
mirrors:
  docker.io:
    endpoint:
      - "https://myregistry.local:5000"
configs:
  "myregistry.local:5000":
    tls:
      insecure_skip_verify: true
```
6. Restart rancher-system-agent. `sudo systemctl restart rancher-system-agent.service`.
7. Check `/etc/rancher/agent/registries.yaml`. It will be `{"configs":{},"mirrors":null}` again.

**Expected behavior**
We can replace `/etc/rancher/agent/registries.yaml` content by cloud-config.

**Environment:**
 - Harvester ISO version: master-9664bf67-head
 - Underlying Infrastructure: KVM
 - Rancher: v2.6.4

**Comment by janeczku** [comment]
@FrankYang0529  When provisioning RKE2 clusters on Harvester you should configure private registry settings in the Cluster options and not by editing the `registries.yaml` file directly.

@guangbochen  This seems like expected behavior and not a bug.

**Comment by FrankYang0529** [comment]
> @FrankYang0529 When provisioning RKE2 clusters on Harvester you should configure private registry settings in the Cluster options and not by editing the `registries.yaml` file directly.

Thanks for your information! I will check whether I can change rancher-system-agent `registries.yaml` by configuring private registry setting.

**Comment by FrankYang0529** [comment]
Verified that we can set a private registry in Cluster Configuration. The initial script uses the private registry in rancher-system-agent instructions. We can see that in rancher-system-agent plan `/var/lib/rancher/agent/applied/xxx.plan`. After rancher-system-agent finishes, it will set the private registry as the `system-default-registry` argument for RKE2. We can see this in `/etc/rancher/rke2/config.yaml.d/50-rancher.yaml`.

![image](https://user-images.githubusercontent.com/4927361/166427115-48a326ab-85d6-4f9a-81b1-972b3caec253.png)

**Comment by FrankYang0529** [comment]
I try again in a non-air-gapped environment. When I set the registry to a random domain, it is not set to rancher-system-agent, but it's set to rke2-server.

From my last [comment](https://github.com/harvester/harvester/issues/2125#issuecomment-1115875641), I think the registry for rancher-system-agent is set to Rancher's one. If we set `systemDefaultRegistry` in Rancher as `myrancher.test:5000`, then rancher-system-agent also use `myrancher.test:5000`.

**Comment by FrankYang0529** [comment]
The issue has been fixed in Rancher v2.6.5-rc6. When I create an RKE2 cluster with a private registry, I can see that rancher-system-agent and RKE2 use this registry to pull images. Close the issue since we don't have any fix from the harvester side.

https://github.com/rancher/rancher/issues/37221#issuecomment-1118033127
